wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/bootstrap/AbstractBootstrapper.java

```
@@ -86,6 +86,9 @@ public void bootstrap() throws InfrastructureException {
         if (finishEventFuture != null) {
             throw new IllegalStateException("Bootstrap method must be called only once.");
         }
+        if (!hasInstallersToBootstrap()) {
```

WDYT about moving such logic to InternalRuntime and do not bootstrap at all if installers are missed

---

Bootstrapper works with each machine, not the whole runtime, InternalRuntime knows nothing about bootstrapper, the bootstrapper is not part of SPI contract. So I do think it is better to not introduce bootstrapper in InternalRuntime

---

I mean changes in DockerInternalRuntime
```
List<Intaller> installers = machineCfg.getInstallers();
if (!installers.isEmpty()) {
    bootstrapperFactory.create(name, getContext().getIdentity(), installers, machine).bootstrap();
}
```

---

i would do both
1. change the contract of the bootstrapper, but rather to throw exception when there is nothing to install, instead of silent quit. What if we do it directly in AbstractBootstrapper constructor? So the contract is that bootstrapper cannot be created with empty list of installers.
2. change the code in internal runtimes(like Sergii suggests) so bootstrapper is not created at all.

---

I like Sergii's idea and not sure about throwing an exception in the case of an empty list. I'll go with his suggestion.